### PR TITLE
Batched Packet Sending

### PR DESCRIPTION
**Pull request by dr7ana**
- packets batched in groups of 8 to reduce syscall overhead
- packet send logic moved to Endpoint from Connection

### CODE REVIEW

File: include/quic/connection.hpp [hunk 1]
```diff
@@ -34,8 +34,10 @@ namespace oxen::quic
 
         int init(ngtcp2_settings& settings, ngtcp2_transport_params& params, ngtcp2_callbacks& callbacks);
 
-        std::array<std::byte, NGTCP2_MAX_UDP_PAYLOAD_SIZE> send_buffer{};
+        static constexpr size_t batch_size = 8;
+        std::array<std::pair<std::array<std::byte, NGTCP2_MAX_PMTUD_UDP_PAYLOAD_SIZE>, size_t>, batch_size> send_buffer{};
```
Comment (jagerman) [inline include/quic/connection.hpp]: This type is unwieldy enough (and used in multiple places) that it ought to be a `using send_buffer_t = std::array<...` typedef.

File: src/connection.cpp [hunk 2]
```diff
@@ -488,12 +487,16 @@ namespace oxen::quic
...
             }
 
-            log::info(log_cat, "Sending data packet with non-stream data frames");
-            if (auto rv = send_packet(nwrite); rv != 0)
-                return;
-            ngtcp2_conn_update_pkt_tx_time(conn.get(), ts);
+            buf.second = nwrite;
+
+            if (++n_packets == batch_size)
+                break;
```
Comment (jagerman) [inline src/connection.cpp]: This doesn't seem right: this `for(;;)` loop may want to send more than 1 packet, but this loop could be restricted to just one iteration (if the previous loop filled up 7/8 slots).  I think we need the same append-and-send-if-full logic as in the upper loop.

File: src/endpoint.cpp [hunk 3]
```diff
@@ -257,18 +257,53 @@ namespace oxen::quic
...
 
-        return io_result{0};
+        std::array<uv_buf_t, batch_size> raw_bufs;
+
+        log::info(log_cat, "Sending udp batch to {}:{}...", p.remote.ip.c_str(), p.remote.port);
+
+        char* packet_data_pos = packet_data->data();
+        for (int i = 0; i < n_pkts; ++i)
+        {
+            assert(bufs[i].second > 0);
```
Comment (tewinget) [inline src/endpoint.cpp]:
```suggestion
            assert(buf[i].second > 0);
```

**Comment by tewinget** [comment]
test 004 aborts and 005 fails, but both pass just fine if I rebase my pending `thread_safety` branch on top of this so apart from the suggested change in `endpoint.cpp` I'd say this is good to go.

**Comment by jagerman** [comment]
Closing (because this is embedded into #1, which takes it much further for performance optimizations)

**Pull request closed without merging**